### ISSUE TITLE
Heap issues

### ISSUE THREAD

**Issue by adamtanana** [pwndbg/pwndbg#535]
### Description
Issues with heap command on ubuntu 18.04

### Steps to reproduce

Program
```c
#include <stdio.h>
#include <stdlib.h>

int main() {
    char* a = malloc(0x40);
    free(a);
}
```

Gdb session history:
```
b main
run
n
n
heap
bins
```

When printing bins it prints incorrect values (there should be one thing in fast bins) and also prints an error
`Error fetching tcache. GDB cannot access thread-local variables unless you compile with -lpthread.
Exception occured: tcachebins: Cannot access memory at address 0x40 (<class 'gdb.MemoryError'>)
For more info invoke `set exception-verbose on` and rerun the command
or debug it by yourself with `set exception-debugger on`

### My setup

Platform: Linux-4.15.0-34-generic-x86_64-with-Ubuntu-18.04-bionic
Gdb:      8.1.0.20180409-git
Python:   3.6.6 (default, Sep 12 2018, 18:26:19)  [GCC 8.0.1 20180414 (experimental) [trunk revision 259383]]
Pwndbg:   1.1.0 build: b847f1d
Capstone: 4.0.1024
Unicorn:  1.0.1

**Comment by GrosQuildu** [comment]
Can't reproduce. Can you provide libc version (`ldd --version`) and full stacktrace (with `set exception-verbose on`)?

**Comment by adamtanana** [comment]
Stack trace:
```
pwndbg> bins
Error fetching tcache. GDB cannot access thread-local variables unless you compile with -lpthread.
Traceback (most recent call last):
  File "/home/adam/pwndbg/pwndbg/commands/__init__.py", line 109, in __call__
    return self.function(*args, **kwargs)
  File "/home/adam/pwndbg/pwndbg/commands/__init__.py", line 200, in _OnlyWhenRunning
    return function(*a, **kw)
  File "/home/adam/pwndbg/pwndbg/commands/__init__.py", line 209, in _OnlyWhenHeapIsInitialized
    return function(*a, **kw)
  File "/home/adam/pwndbg/pwndbg/commands/heap.py", line 335, in tcachebins
    tcachebins = main_heap.tcachebins(addr)
  File "/home/adam/pwndbg/pwndbg/heap/ptmalloc.py", line 434, in tcachebins
    count = int(counts[i])
  File "/home/adam/pwndbg/pwndbg/inthook.py", line 42, in __new__
    value = value.cast(pwndbg.typeinfo.long)
gdb.MemoryError: Cannot access memory at address 0x40

If that is an issue, you can report it on https://github.com/pwndbg/pwndbg/issues
(Please don't forget to search if it hasn't been reported before)
To generate the report and open a browser, you may run `bugreport --run-browser`
PS: Pull requests are welcome
fastbins
0x20: 0x0
0x30: 0x0
0x40: 0x0
0x50: 0x0
0x60: 0x0
0x70: 0x0
0x80: 0x0
unsortedbin
all: 0x0
smallbins
empty
largebins
empty
pwndbg> 
```

My ldd version is:
```
ldd (Ubuntu GLIBC 2.27-3ubuntu1) 2.27
```

**Comment by disconnect3d** [comment]
@adamtanana It is also worth to mention the compilation flags you used. I assume you used `-g` or `-ggdb` as you use `n` command (or `next`) which jumps over a source code instruction instead of assembly instruction.

@GrosQuildu This might be the reason why you couldn't reproduce.

---

Below is my log.
```
pwndbg> version
Gdb:      8.1.0.20180409-git
Python:   3.6.5 (default, Apr  1 2018, 05:46:30)  [GCC 7.3.0]
Pwndbg:   1.1.0 build: b847f1d
Capstone: 4.0.1024
Unicorn:  1.0.1
```

The position where I am after running the commands given in the issue:
```
───────────────────────────────────────────────────────────────────────────────────[ SOURCE (CODE) ]───────────────────────────────────────────────────────────────────────────────────
In file: /home/dc/playground/pwndbg_heap_problem/main.c
   2 #include <stdlib.h>
   3 
   4 int main() {
   5     char* a = malloc(0x40);
   6     free(a);
 ► 7 }
```
or in asm:
```
   0x5555555546ac <main+34>                  mov    eax, 0
 ► 0x5555555546b1 <main+39>                  leave  
   0x5555555546b2 <main+40>                  ret    
    ↓
```

The result of `heap` and `bins`:

```
pwndbg> heap
0x555555756000 PREV_INUSE {
  mchunk_prev_size = 0, 
  mchunk_size = 593, 
  fd = 0x1000000, 
  bk = 0x0, 
  fd_nextsize = 0x0, 
  bk_nextsize = 0x0
}
0x555555756250 FASTBIN {
  mchunk_prev_size = 0, 
  mchunk_size = 81, 
  fd = 0x0, 
  bk = 0x0, 
  fd_nextsize = 0x0, 
  bk_nextsize = 0x0
}
0x5555557562a0 PREV_INUSE {
  mchunk_prev_size = 0, 
  mchunk_size = 134497, 
  fd = 0x0, 
  bk = 0x0, 
  fd_nextsize = 0x0, 
  bk_nextsize = 0x0
}
pwndbg> bins
Error fetching tcache. GDB cannot access thread-local variables unless you compile with -lpthread.
Exception occured: tcachebins: Cannot access memory at address 0x40 (<class 'gdb.MemoryError'>)
For more info invoke `set exception-verbose on` and rerun the command
or debug it by yourself with `set exception-debugger on`
fastbins
0x20: 0x0
0x30: 0x0
0x40: 0x0
0x50: 0x0
0x60: 0x0
0x70: 0x0
0x80: 0x0
unsortedbin
all: 0x0
smallbins
empty
largebins
empty
```

**Comment by disconnect3d** [comment]
I also believe that this:
```
Error fetching tcache. GDB cannot access thread-local variables unless you compile with -lpthread.
```
is a valid response, but it is not good - from an UX point of view - that we show an exception too.

**Comment by adamtanana** [comment]
Compiled with 
> gcc test.c -g -lpthread
> gcc test.c -g

Both show same output
`Error fetching tcache. GDB cannot access thread-local variables unless you compile with -lpthread.`)

On another note I just noticed that when running `fastbins` w/out errors, it should show one chunk in the fastbin list shouldn't it?
> malloc(0x40) <= is a fast bin size?

However it shows the empty list
```
pwndbg> fastbins
fastbins
0x20: 0x0
0x30: 0x0
0x40: 0x0
0x50: 0x0
0x60: 0x0
0x70: 0x0
0x80: 0x0
```

**Comment by adamtanana** [comment]
My colleague running the following versions gets the correct output however
```
pwndbg> bins
fastbins
0x20: 0x0
0x30: 0x0
0x40: 0x0
0x50: 0x602000 ◂— 0x0
0x60: 0x0
0x70: 0x0
0x80: 0x0
unsortedbin
all: 0x0
smallbins
empty
largebins
empty
pwndbg> Quit
pwndbg> version
Gdb:      7.11.1
Python:   3.5.2 (default, Nov 23 2017, 16:37:01)  [GCC 5.4.0 20160609]
Pwndbg:   1.1.0 build: b322d3a
Capstone: 4.0.1024
Unicorn:  1.0.1
pwndbg> 
```

**Comment by GrosQuildu** [comment]
In glibc 2.27 fastbin chunk goes to tcachebin first, so doesn't appear in fastbin list. So that part is correct (afaik, may be wrong ;) ). Your's colleague must use older glibc version.

As for tcache error, @disconnect3d corretly pointed that the error is valid, gdb can't find info about tcache, check with:
```
pwndbg> p tcache
Cannot find thread-local storage for process 12377, shared library libc.so.6:
Cannot find thread-local variables on this target
```
Recompile binary with `-lpthread`, then in gdb run this [two commands](https://sourceware.org/glibc/wiki/Testing/Builds):
```
pwndbg> set libthread-db-search-path /path/glibc_versions/2.27/nptl_db/
pwndbg> set auto-load safe-path /path/glibc_versions/2.27/nptl_db/
pwndbg> start
```
Tell if this helped.

About misleading exception, gdb returns strange info with this command:
```
pwndbg> info address tcache
Symbol "tcache" is a thread-local variable at offset 0x40 in the thread-local storage for `/path/2.27/libc.so.6'.
```
0x40 value is returned [here](https://github.com/pwndbg/pwndbg/blob/dev/pwndbg/heap/ptmalloc.py#L144). Then in line 148 exception occurs, but `self._thread_cache` is set and then is resolved resulting in memory exception. It is strange...

 To fix that we may add `self._thread_cache = None` after printing error (after line 152).

Moreover, `tcache` returned wrong info once I get it working:
```
pwndbg> tcache
{
  counts = "\020`uUUU\000\000\000\000\000\000\000\000\000\000"..., 
  entries = {0x0, 0x0, 0x7ffff7ff3740, 0x7ffff7ff40a0, 0x7ffff7ff3740, 0x0, 0x0, 0x5bc113ad9ffaa700, 0x83af3831bce42024, 0x0 <repeats 55 times>}
}
pwndbg> p *tcache
$1 = {
  counts = "\000\000\000\001", '\000' <repeats 59 times>, 
  entries = {0x0, 0x0, 0x0, 0x555555756260, 0x0 <repeats 60 times>}
}

pwndbg> p &tcache
$2 = (tcache_perthread_struct **) 0x7ffff7ff36f0
pwndbg> x/xg 0x7ffff7ff36f0
0x7ffff7ff36f0:	0x0000555555756010
pwndbg> p *(tcache_perthread_struct*)0x0000555555756010
$3 = {
  counts = "\000\000\000\001", '\000' <repeats 59 times>, 
  entries = {0x0, 0x0, 0x0, 0x555555756260, 0x0 <repeats 60 times>}
}
```

`tcache_addr` [here](https://github.com/pwndbg/pwndbg/blob/dev/pwndbg/heap/ptmalloc.py#L144) is 0x7ffff7ff36f0, so it should be dereferenced one more time?

Help of somebody more into tcache would be useful ;)

**Comment by disconnect3d** [comment]
@Segflow @mdebski @andigena can you guys help here?

**Comment by disconnect3d** [comment]
Hey,

@Segflow @mdebski @andigena 

I have fixed some bugs regarding tcache in https://github.com/pwndbg/pwndbg/pull/552 and added a its address retrieval when binary is compiled w/o `-lpthread`. If someone could look on it, check if it works on your machine and make a code review, I would be pleased! :)

**Comment by disconnect3d** [comment]
I believe that:
* the missing fastbin is an invalid bug - the bin went to tcache instead
* the tcache bug has been fixed with #552 - and we also try to display it when binary is compiled without `-lpthread`

So I am closing it for now. o/
